src/grammar/expressions.rs

```diff
@@ -90,61 +113,134 @@ fn expr_bp(p: &mut Parser, min_bp: u8) {
             continue;
         }
 
-        if let Some(((l_bp, r_bp), syntax_kind)) = infix_bp(op) {
-            if l_bp < min_bp {
+        if let Some(operator) = infix_bp(op) {
+            if operator.bp.l_bp() < min_bp {
                 break;
             }
 
-            match syntax_kind {
+            match operator.mapping {
                 Some(syntax_kind) => p.bump_any_map(syntax_kind),
                 None => p.bump_any(),
             }
 
-            add_expr_node(p, checkpoint, Some(r_bp));
+            if let Some(cb) = operator.callback {
+                cb(p, operator.bp.r_bp());
+            }
+
+            add_expr_node(p, checkpoint, Some(operator.bp.r_bp()));
             continue;
         }
 
         break;
     }
+    Ok(())
 }
 
 fn add_expr_node(p: &mut Parser, checkpoint: Checkpoint, sub_expr: Option<u8>) {
     p.start_node_at(checkpoint, SyntaxKind::Expression);
 
     if let Some(min_bp) = sub_expr {
-        expr_bp(p, min_bp)
+        let _ = expr_bp(p, min_bp);
     };
 
     p.finish();
 }
 
-fn prefix_bp(op: TokenKind) -> Option<(u8, Option<SyntaxKind>)> {
-    match op {
-        TokenKind::NotKw => Some((5, Some(SyntaxKind::LogicOp))),
-        TokenKind::Plus | TokenKind::Minus => Some((15, None)),
-        _ => None,
+struct BindingPower(u8, u8);
+
+impl BindingPower {
+    pub fn l_bp(&self) -> u8 {
+        self.0
+    }
+
+    pub fn r_bp(&self) -> u8 {
+        self.1
     }
 }
 
-fn postfix_bp(op: TokenKind) -> Option<(u8, Option<SyntaxKind>)> {
-    match op {
-        TokenKind::Exclam => Some((17, None)),
-        _ => None,
+type Callback = &'static dyn Fn(&mut Parser, u8);
+
+struct Operator {
+    bp: BindingPower,
+    mapping: Option<SyntaxKind>,
+    callback: Option<Callback>,
+}
+
+impl Operator {
+    fn ctor(l_bp: u8, r_bp: u8, mapping: Option<SyntaxKind>, callback: Option<Callback>) -> Self {
```

Comment:
Constructors in Rust are named `new` by convention for most situations.

Reply:
The `ctor` method is only being used internally and is not `pub`. Should we rename the existing `new` method?